### PR TITLE
feat: rate limiter + unit tests + readme

### PR DESCRIPTION
**Pull request by fforbeck**
This PR add tests for the Rate Limiter middleware, the main changes include:

- [x] Updated README with steps and configs to run it locally
   - [x] Project Setup
   - [x] Running Tests
- [x] Added Mocha, Chai and Sinon packages to run Unit Tests
- [x] Rate Limiter
   - [x] Feature Flag (env variable)
   - [x] Fixes
      - [x] Rate Limiter middleware needs to be placed after the IPFS Parsed URL middleware, so we can access the CID 
      - [x] Refactored the `import {version} from './package.json' to be set via task and imported as ENV variable, so we can easily tweak that in the tests and don't run into import issues
   - [x] Unit Tests
     - 1. Calls next handler if no auth token is provided and rate limit is not exceeded
     - 2. Throws an error if no auth token is provided and rate limit is exceeded
     - 3. Calls next handler if auth token is present but no token metadata is found and rate limit is not exceeded
     - 4. Throws an error if auth token is present but no token metadata is found and rate limit is exceeded
     - 5. Calls next handler if auth token is present and token metadata is invalid but rate limit is not exceeded
     - 6. Throws an error if auth token is present and token metadata is invalid and rate limit is exceeded
     - 7. Calls next handler if auth token is present and token metadata is valid
   - [x] Update Github Action to execute Unit Tests

### CODE REVIEW

**Comment by travis** [comment]
Could you say more about why Jest is useful here? We've typically used `node:test` for this project, and Mocha (https://mochajs.org/) or AVA (https://github.com/avajs/ava) on other projects - I'm not entirely sure why we've shied away from Jest, but I am a bit hesitant to add yet another testing framework to the mix for our team. Could definitely be convinced - I've used Jest in the past and it's generally pretty good - but need a bit more convincing ;-p

**Comment by fforbeck** [comment]
> Could you say more about why Jest is useful here? We've typically used `node:test` for this project, and Mocha (https://mochajs.org/) or AVA (https://github.com/avajs/ava) on other projects - I'm not entirely sure why we've shied away from Jest, but I am a bit hesitant to add yet another testing framework to the mix for our team. Could definitely be convinced - I've used Jest in the past and it's generally pretty good - but need a bit more convincing ;-p

sure thing @travis 

I don’t have super strong opinions on this either, but Jest does give you some nice built-in features like mocking, spies, and coverage reports right out of the box. With Mocha (or node:test), you’d need to add additional libraries like Sinon for mocks and nyc for coverage, which Jest handles natively.

That said, if we can use Mocha alongside node:test, it should work fine, even though Mocha isn't as feature-rich in those areas. Totally open to whichever approach the team prefers! 😊

**Comment by travis** [comment]
ok cool - I lean toward Mocha, but tbh I'd like @alanshaw to weigh in since he's written a lot more of this code (like, infinitely more, as this is really my first freeway contribution as well!) - Alan any preferences/thoughts on testing frameworks here?

File: package.json
Comment (travis) [inline package.json:19]: question: should these just be added to `prepare` above? it would be nice to avoid duplicating this line, and seems like `set-version` can probably safely be run any time `prepare` is triggered - does that sound right to you?

**Review by travis — CHANGES_REQUESTED** [review]
I've added a few comments, some of them blocking - I took https://conventionalcomments.org/ for a spin on this review, curious if you like it! 

I probably won't Approve this PR in general since I've never contributed to Freeway myself - @alanshaw could you take a look at this when you get a chance?

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:53]:
```suggestion
    console.log(`rate limit exceeded: ${cid}`)
```

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:71]: Maybe pass the cache and accounting API instace to this function? In general we should avoid passing everything everywhere since it makes it harder to test.

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:43]: This only ever uses one value from env - consider simply passing the API instance not the entire environment.

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:74]: Please take a look at: https://developers.cloudflare.com/workers/runtime-apis/context/#waituntil

Do we need to wait for this as part of the request?

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:87]: Please can we have some type information here?

File: src/bindings.d.ts
Comment (alanshaw) [inline src/bindings.d.ts:36]: I think we don't generally define types for like this as we'd never pass this around in code. Instead this would just be a file with an exported function called create.

File: src/bindings.d.ts
Comment (alanshaw) [inline src/bindings.d.ts:27]: `RateLimitExceeded` - I don't see this defined anywhere...

Comment (alanshaw) [inline src/bindings.d.ts:27]: `options: GetCIDRequestOptions` are they optional? If not then specify as `config` or `context`.

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:150]: Log just once at startup please not every request!

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:126]: Yes eventually this will be indexing service, but you may have to use content claims temporarily. Either way I believe the current idea is for them to have the same API.

**Review by alanshaw — CHANGES_REQUESTED** [review]
Love the docs updates 🙏 thank you.

File: src/middleware.js
Comment (alanshaw) [inline src/middleware.js:149]:
```suggestion
    if (env.FF_RATE_LIMITER_ENABLED !== 'true') {
```

It'll be a string or undefined. Best to explicitly require the string "true".

**Comment by alanshaw** [comment]
> > Could you say more about why Jest is useful here? We've typically used `node:test` for this project, and Mocha (https://mochajs.org/) or AVA (https://github.com/avajs/ava) on other projects - I'm not entirely sure why we've shied away from Jest, but I am a bit hesitant to add yet another testing framework to the mix for our team. Could definitely be convinced - I've used Jest in the past and it's generally pretty good - but need a bit more convincing ;-p
> 
> sure thing @travis
> 
> I don’t have super strong opinions on this either, but Jest does give you some nice built-in features like mocking, spies, and coverage reports right out of the box. With Mocha (or node:test), you’d need to add additional libraries like Sinon for mocks and nyc for coverage, which Jest handles natively.
> 
> That said, if we can use Mocha alongside node:test, it should work fine, even though Mocha isn't as feature-rich in those areas. Totally open to whichever approach the team prefers! 😊

Slightly my bad here, I tried `node --test` because it was new, but turns out it isn't great so I'd welcome a switch to something else. We also don't use it in any other repo.

I'm mildly against having 2 testing frameworks in the same repo, but at the same time I don't want to waste time here - there's a bunch of tests written now using it so...meh. The point is more that testing exists at all - I don't mind which framework we use. People like consistency though, and using the same framework(s) everywhere allows us to work more quickly and without context switches. We should establish a single/set of frameworks we're comfortable using as a team. I realise that may have changed now. @travis this is something needed to be decided for team norms.

Just a couple of notes: 1) I though jest was legacy/unmaintained? 2) We don't use jest _anywhere_ else.

File: package.json
Comment (fforbeck) [reply package.json:19]: Yes, that sounds right. I will update that. thanks

Comment (fforbeck) [reply package.json:19]: ah hold on, prepare task is run when you install the packages, but we can improve this code for sure.

File: set-version-env.js
Comment (fforbeck) [reply set-version-env.js:13]: It is missing the `process.env.VERSION=version` - I will add it. The idea is to use this script to read the version from package json and set it into the `VERSION` env var instead of importing the `package.json` file directly in the middleware.js - which was causing several failures in the tests because it couldn't resolve the `.json` import. The version is used by the middleware that returns the API version.

File: src/bindings.d.ts
Comment (fforbeck) [reply src/bindings.d.ts:36]: I think it will be better if we use separate files for each middleware/handler instead of defining everything in `middleware.js`, otherwise it will be hard to maintain all of them in one place. Then I can refactor this as well.

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js:22]: Agree. We already have the `handlers` directory, I will use that to define the rate-limit.js middleware.

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js:71]: The issue I see with passing the cache and accounting to this function is that we will always need to call the `Accounting.create...` to provide that instance in the function call, and if the value is in the cache already we don't even need to execute that call at all because it just decodes the cached value.

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js:74]: Good catch. Yes, I believe we will need to use the `ctx.waitUntil` call here.

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js:149]: It is typed as a boolean in the Environment bindings.d.ts

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js:251]: My bad. Fixed the type declaration and now we don't even need that conditional.

File: src/middleware.js
Comment (Peeja) [inline src/middleware.js:67]: **suggestion:** Let's not leave this as a `TODO` here; let's capture the user value and move it to a card we can prioritize on the board. We can't decide to actually do this until it's in the flow and context of our other work.

File: src/middleware.js
Comment (Peeja) [inline src/middleware.js:74]: **suggestion:** Should we add an `expirationTtl` on this? Right now (especially given we're not SWRing), we appear to cache these indefinitely.

File: src/middleware.js
Comment (Peeja) [inline src/middleware.js:113]: **suggestion:** Untangle this a bit so we have the one path to `RATE_LIMIT_EXCEEDED.NO` followed by one fallback to `checkRateLimitForCID()`.

```suggestion
const RateLimits = {
  create: ({ env }) => ({
    check: async (cid, options) => {
      // We'll serve the request without a rate limit if...
      const authToken = await getAuthorizationTokenFromRequest(options)

      // ...there's an auth token in the request...
      if (authToken) {
        console.log(`found token ${authToken}, looking for content commitment`)
        const tokenMetadata = await getTokenMetadata(env, authToken)

        // ...and we can find a content commitment for the token...
        if (tokenMetadata) {
          // ...and the token is valid.
          if (!tokenMetadata.invalid) {
            return RATE_LIMIT_EXCEEDED.NO
          }
        }
      }

      // Otherwise, we'll rate limit the request.
      return checkRateLimitForCID(env, cid)
    }
  })
}
```

File: src/bindings.d.ts
Comment (Peeja) [inline src/bindings.d.ts:26]: **suggestion:** This type didn't exist yet, so it was `any`.

```suggestion
export type RateLimitExceeded = typeof RATE_LIMIT_EXCEEDED[keyof typeof RATE_LIMIT_EXCEEDED]

```

File: src/middleware.js
Comment (Peeja) [inline src/middleware.js:63]: **suggestion:** Actually type this. 😜

```suggestion
 * @returns {Promise<import('./bindings.js').TokenMetadata | null>}
```

File: src/middleware.js
Comment (Peeja) [inline src/middleware.js:160]: **suggestion:** I think we don't need to—we do get some basic HTTP logging for free on Cloudflare, right?

**Review by Peeja — CHANGES_REQUESTED** [review]

**Comment by Peeja** [comment]
FWIW, I'm highly in favor of Jest. It's not legacy; quite the opposite, AFAIK it's where the community primarily is right now. Granted I'm biased by being used to Jest, but I'd love to decide to (in the long run) move _everything_ to Jest.

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js:40]: @alanshaw the issue is that `IpfsUrlContext` returns the `dataCid` as type `multiformats/cid`. Should we use the same?

File: wrangler.toml
Comment (fforbeck) [reply wrangler.toml:1]: Yes, that's right!

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js:139]: Makes sense. I will change that. Thanks

File: README.md
Comment (fforbeck) [reply README.md:60]: I think having them in the Readme should be fine for now. That `dev` script starts the worker in dev mode (but that was for miniflare v2), I've updated it to miniflare v3 and I guess we can remove that `dev` script for now because they recommend using `npx wrangler dev -e <profile>` instead.

File: .github/actions/test/action.yml
Comment (fforbeck) [reply .github/actions/test/action.yml]: Yeah. Good point. I will change that.

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js]: Thanks for the suggestion. I leave this to be changed in another PR.

File: src/middleware.js
Comment (alanshaw) [reply src/middleware.js]: Hmm, yes I see but that doesn't automatically turn it into a boolean - it's just an environment variable right, either set in `wrangler.toml` or as a deployment variable - it will be a string unless you explicitly cast it in code.

File: set-version-env.js
Comment (alanshaw) [reply set-version-env.js]: Oh ok got it - because new tests are running against actual source code not esbuild bundle.

Please next time just leave a comment in the PR description to explain! 🙏

File: src/middleware.js
Comment (alanshaw) [reply src/middleware.js]: ACK yes agree. We should throw an error if rate limter is not configured then.

File: src/middleware.js
Comment (alanshaw) [reply src/middleware.js]: Oh, yes I guess it is old, I'd use `import('multiformats').CID` instead here then.

File: src/handlers/rate-limiter.js
Comment (alanshaw) [inline src/handlers/rate-limiter.js:121]: You can `get` and `put` and JSON decode/dnecode so you perhaps don't need these encode/decode functions?:

https://developers.cloudflare.com/kv/api/read-key-value-pairs/#get-method

**Review by alanshaw — CHANGES_REQUESTED** [review]

File: src/middleware.js
Comment (alanshaw) [reply src/middleware.js]: Yes, understood but I'm not convinced it'll be a performance problem.

Related, I'd probably implement the cache as a transparent wrapper around the accounting service and pass the accounting service as a dependency to the rate limiting service.

i.e. something like:

```js
const accounting = Accounting.create({ serviceURL: env.ACCOUNTING_SERVICE_URL })
const cachedAccounting = Accounting.withCache(accounting, env.AUTH_TOKEN_METADATA)
const rateLimiter = RateLimiter.create(cachedAccounting)

const status = await rateLimiter.check(cid, request)
if (status  === RATE_LIMIT_EXCEEDED.YES) {
  // ...
```

So then it should be easier to test the rate limiter, since the rate limiter is not importing a function that requires a KV _and_ an accounting service. Also easier to test the caching logic in isolation.

...anyways...just a suggestion - non blocking.

File: src/middleware.js
Comment (fforbeck) [reply src/middleware.js]: Oh ok. That makes sense. Thanks for the clarification.

File: src/handlers/rate-limiter.js
Comment (fforbeck) [reply src/handlers/rate-limiter.js:121]: @alanshaw The current API doesn't support that `json` type. I won't change this now because this PR is getting too big, but will certainly update the cloudflare packages in another PR to use the latest version available - which supports `json` values.

File: src/handlers/rate-limiter.js
Comment (fforbeck) [reply src/handlers/rate-limiter.js:116]: We do use that to call `ctx.waitUntil` at [line 128]( https://github.com/storacha/freeway/pull/115/files/e37c45c88caebca3f35314cb9d8a5b98c303baf5#diff-29338ef628c3debc8fcd22bcfa9b84b082aa8be54c8c372bb071537b234159c3R128).

**Comment by fforbeck** [comment]
@alanshaw I've applied pretty much all the suggestions, however, I left some of them to be done in different PRs to minimize the number of changes here.
TODO
- Rename URLs from Readme
- Implement the cache as a transparent wrapper around the accounting service and pass the accounting service as a dependency on the rate-limiting service. https://github.com/storacha/freeway/pull/115#discussion_r1784309869
- Log "rate limit disabled" just once at startup please not every request
- Update Cloudflare workers library to use the latest KV `put/get` types - which support JSON values

File: src/bindings.d.ts
Comment (alanshaw) [inline src/bindings.d.ts:36]: Minor, but the same with this one - we'd typically not define this in types and just define it in the file and use like:

```js
import * as Accounting from 'services/accounting.js'

Accounting.create(...)
```

...and in `accounting.js`

```js
export const create = () => { /* ... */ }
```

**Review by alanshaw — APPROVED** [review]

File: src/bindings.d.ts
Comment (fforbeck) [reply src/bindings.d.ts:36]: Sounds good. I will add it to the TODO list as well. Thanks

**Review by travis — APPROVED** [review]

**Pull request merged**
